In [2]:
s = list("ZUNEH TDTHR IFTAD JBFCA TWHHZ TSKXK ZRPED BCTIX TBVGJ KDGVS JJIWF NJUPS UDSCN TETXV CYVCH VEFJH JUDSK DYYTS VMYXT SVGFK XTEHT WRWVP YLGJJ LMRIX KPYVB JEIBF JQURT EIFZC YYTRF HYZCK FGRRI NFCNE IMVUJ NTXKL TISXZ QJCXJ MTNKX XKWJR ITDXH YNUFI MVHNJ DWKWJ RITDX HWPHK DWNWF KTAVG DFJBZ HMKDH RAQZI YYPYR AQKWN EVXRG JDPIV DKRIT DHQZI YCTUR GYZRQ VHYYP YDDAV PWFJS UXSGT WGTYL PQDDY ZDSRI YIPHK XSXTF TWTKW JILMV CYYTD RGJRA NKIQV SNJIF ERJRE FIIGL IWVEJ CANEV ZGDSS TNEVX HJJVO JUITX TYYTW ZCYYP YFCJJ TSKTS TTDFJ BZAQJ TJKWJ ITNJP SVCTI BTLHF DDZEI TWXSW DWDPY ZDSRQ TLIYY TBFGQ UXKAJ XKPQZ IYCTN DPLZC FKXTE PSUIM ZCPZC LRGJR EUCXJ U".replace(" ",""))
ss = "ZUNEH TDTHR IFTAD JBFCA TWHHZ TSKXK ZRPED BCTIX TBVGJ KDGVS JJIWF NJUPS UDSCN TETXV CYVCH VEFJH JUDSK DYYTS VMYXT SVGFK XTEHT WRWVP YLGJJ LMRIX KPYVB JEIBF JQURT EIFZC YYTRF HYZCK FGRRI NFCNE IMVUJ NTXKL TISXZ QJCXJ MTNKX XKWJR ITDXH YNUFI MVHNJ DWKWJ RITDX HWPHK DWNWF KTAVG DFJBZ HMKDH RAQZI YYPYR AQKWN EVXRG JDPIV DKRIT DHQZI YCTUR GYZRQ VHYYP YDDAV PWFJS UXSGT WGTYL PQDDY ZDSRI YIPHK XSXTF TWTKW JILMV CYYTD RGJRA NKIQV SNJIF ERJRE FIIGL IWVEJ CANEV ZGDSS TNEVX HJJVO JUITX TYYTW ZCYYP YFCJJ TSKTS TTDFJ BZAQJ TJKWJ ITNJP SVCTI BTLHF DDZEI TWXSW DWDPY ZDSRQ TLIYY TBFGQ UXKAJ XKPQZ IYCTN DPLZC FKXTE PSUIM ZCPZC LRGJR EUCXJ U".replace(" ", "")

# Home Work 3 Problem 1 -- Tony Schleck

In [3]:
# Finding the maximum number of n-grams that appear in the cipher. n-grams find the most used strings in a text, For example the most common 2-grams in English are "TH" (1.5%), "HE" (1.3%), and "IN" (0.9%). 
from nltk.util import ngrams
from collections import Counter
max_ngram = 5
minimum_count = 3
ngrams_found = Counter()

for x in range(max_ngram-1):
    ngrams_found += Counter(["".join(ngram) for ngram in ngrams(ss, x+minimum_count)])
for key, val in ngrams_found.items():
    if len(key) == 5 and val > 1:
        print(key, val)

FKXTE 2
KWJRI 2
WJRIT 2
JRITD 2
RITDX 2
ITDXH 2
DFJBZ 2
QZIYC 2
ZIYCT 2
YZDSR 2


In [4]:
differences = []
w = ss.rfind('FKXTE')
z = ss.find('FKXTE')

differences.append(w-z)

In [5]:
w = ss.rfind('KWJRI')
z = ss.find('KWJRI')

differences.append(w-z)

In [6]:

w = ss.rfind('DFJBZ')
z = ss.find('DFJBZ')

differences.append(w-z)

In [7]:
w = ss.rfind('QZIYC')
z = ss.find('QZIYC')

differences.append(w-z)

In [8]:
w = ss.rfind('YZDSR')
z = ss.find('YZDSR')

differences.append(w-z)

In [9]:
differences

[402, 21, 192, 216, 150]

We can find the Greatest Common Factor of the differences to find our period m. 
The greatest common factor of 402,21,192,216 and 150 is 3. So our period is 3, meaning the cipher repeats every 3 Letters. 

In [10]:
#splitting our string into 3 decimations:
s1=""
s2=""
s3=""

for i in range(len(s)):
    ch = s[i]
    if i%3 == 0:
        s1 += ch
    if i%3 == 1:
        s2 += ch
    else:
        s3+=ch
l = [s1,s2,s3]

Next we will find the index of coincidence for each decimation (every 3 letters shold have an index of coincidence close to the Egnlish language 0.065) 


In [11]:
import collections
from collections import Counter

def IC(s):
    f = dict()
    f = dict(Counter(''.join(s)))
    od = collections.OrderedDict(sorted(f.items()))
    
    sum_IC = 0
    N = len(s)
    
    for key in od:
        c = od[key]
        sum_IC += ( (c*(c-1)) / (N*(N-1)) )
        
    return round(sum_IC, 3)

tot = 0
for i in range(3):
    print("The index of coincidence for decimation", i+1, "is: ", IC(l[i]))
    tot+=IC(l[i])
print("The average IC is: ",tot/3)

The index of coincidence for decimation 1 is:  0.067
The index of coincidence for decimation 2 is:  0.074
The index of coincidence for decimation 3 is:  0.05
The average IC is:  0.06366666666666666


Our average index of coincidence gives us additional evidence that the period is 3 because the index of coincidence is very close to the index of coincidence for regular english is around 0.06

Next we will sort the cipher letters into groups of every three letters. these three groups of letters had the same affine cipher applied per group. 

In [12]:
f = dict()
f = dict(Counter(''.join(ss)))
od = collections.OrderedDict(sorted(f.items()))

In [13]:
odl = list()
for i in range(3):
    f = dict()
    f = dict(Counter(''.join(l[i])))
    od = collections.OrderedDict(sorted(f.items()))
    odl += [od]

In [15]:
#Displaying the letter frequency of each decimation 
from IPython.core.display import display, HTML
tables = ["","",""]
html_string = "<table><tr><th>Cipher Letter</th><th>Count</th><th>Frequency</th></tr>"
c=0
for od in odl:
    html_string = "<table><tr><th>Cipher Letter</th><th>Count</th><th>Frequency</th></tr>"
    for key in od:
        new = "<tr><td>" + key + "</td><td>" + str(od[key]) + "</td><td>" + str(round(od[key]/len(s1), 3))+"</td></tr>"
        html_string += new
    html_string += "</table>"
    tables[c] = html_string
    c+=1


for table in tables:
    display(HTML(table))

In [16]:


import numpy as np

cn = np.array([ord(x) - 65 for x in ss])
cn = list(cn)

for i in range(len(cn)):
    #R
    if i%3 == 0:
        cn[i]-= 17
    #P    
    elif i%3 == 1:
        cn[i]-= 41
    #F    
    else:
        cn[i]-=57
    
    cn[i] = cn[i]%26
#KEY: RPF

In [17]:
for num in range(len(cn)):
    cn[num] = (cn[num])+65

In [18]:
y=cn
list_string = [chr(x) for x in y]
string = ""
dec = string.join(list(list_string))
dec

'IFINSOMECATACLYSMALLOFSCIENTIFICKNOWLEDGEWERETOBEDESTROYEDANDONLYONESENTENCEPASSEDONTOTHENEXTGENERATIONSOFCREATURESWHATSTATEMENTWOULDCONTAINTHEMOSTINFORMATIONINTHEFEWESTWORDSIBELIEVEITISTHEATOMICHYPOTHESISORTHEATOMICFACTORWHATEVERYOUWISHTOCALLITTHATALLTHINGSAREMADEOFATOMSLITTLEPARTICLESTHATMOVEAROUNDINPERPETUALMOTIONATTRACTINGEACHOTHERWHENTHEYAREALITTLEDISTANCEAPARTBUTREPELLINGUPONBEINGSQUEEZEDTOGETHERINTHATONESENTENCEYOUWILLSEETHEREISANENORMOUSAMOUNTOFINFORMATIONABOUTTHEWORLDIFJUSTALITTLEIMAGINATIONANDTHINKINGAREAPPLIED'

The final text with correct spacing reads: "IF IN SOME CATACLYSMALL OF SCIENTIFIC KNOWLEDGE WERE TO BE DESTROYED AND ONLY ONE SENTENCE PASSED ON TO THE NEXT GENERATIONS OF CREATURES WHAT STATEMENT WOULD CONTAIN THE MOST INFORMATION IN THE FEWEST WORDS I BELIEVE IT IS THE ATOMIC HYPOTHESIS OR THE ATOMIC FACTOR WHATEVER YOU WISH TO CALL IT THAT ALL THINGS ARE MADE OF ATOMS LITTLE PARTICLES THAT MOVE AROUND IN PERPETUAL MOTION ATTRACTING EACH OTHER WHEN THEY ARE A LITTLE DISTANCE APART BUT REPELLING UPON BEING SQUEEZED TOGETHER IN THAT ONE SENTENCE YOU WILL SEE THERE IS AN ENORMOUS AMOUNT OF INFORMATION ABOUT THE WORLD IF JUST A LITTLE IMAGINATION AND THINKING ARE APPLIED" 